### PR TITLE
More _from metadata fixes for shrinkwrap reliability (#11736 and #12347)

### PR DESCRIPTION
**Pull request by sminnee**
One of these @iarna has already mentioned in comments in #12347.

These fixes both related to improving `npm shrinkwrap` reliability by getting more a consistent definition of `_from` values.

It still needs tests, which I'll try and get to over the next few days, but I wanted to expose my work-to-date.

### CODE REVIEW

**Comment by sminnee** [comment]
Note that #11373 / #11493 is a related issue still not resolved by this.

File: lib/cache/add-remote-git.js [hunk 1]
```diff
@@ -375,14 +383,6 @@ function updateSubmodules (from, resolvedURL, tmpdir, cb) {
         // https://github.com/npm/npm/issues/6400
         addLocal(spec, null, function (er, data) {
           if (data) {
-            if (npm.config.get('save-exact')) {
-              log.verbose('addRemoteGit', 'data._from:', resolvedURL, '(save-exact)')
```
Comment (iarna) [inline lib/cache/add-remote-git.js]: I think even with your patch we need this behavior if `--save-exact` was used, as it saves with a commit id.

**Comment by iarna** [comment]
Status note: This is waiting on an update that restores `--save-exact` behavior.

**Comment by xdissent** [comment]
Anything else I can do to help with this one?

**Review by zkat — CHANGES_REQUESTED** [review]
Going to make a note also that this code literally does not exist anymore.

I'm not sure off the top of my head what the corresponding patch would be, though, and whether pacote would need to be patched -- pacote shouldn't have this sort of awareness about saving, imo.

**Comment by iarna** [comment]
As the code this PR is patching is no longer in use (and indeed, I don't believe this is still an issue) I'm going to close this.

The original issue here, IIRC, is that subsequent installs from a shrinkwrap would end up with a `_from` pointing at the tarball URL and then that would go into the `from` field in future shrinkwraps, causing churn there.

Values for the shrinkwraps are no longer produced so naively, but more importantly, the `from` field is no longer included in shrinkwraps.

**Pull request closed without merging**